// @favelasquez/agentskills/src/detect.mjs
import { readFileSync, existsSync, readdirSync } from 'node:fs';
import { join, resolve, extname } from 'node:path';
import { homedir } from 'node:os';
import {
  SKILLS_MAP,
  UNIVERSAL_SKILLS,
  FRONTEND_PACKAGES,
  WEB_FRONTEND_EXTENSIONS,
} from './skills-map.mjs';
import { AGENTS } from './agents.mjs';

// ── Constants ────────────────────────────────────────────────

const SKIP_DIRS = new Set([
  'node_modules', '.git', 'vendor', '.next', 'dist', 'build',
  '.output', '.nuxt', '.svelte-kit', '__pycache__', '.cache',
  'coverage', '.turbo', 'var',
]);

// ── File Readers ─────────────────────────────────────────────

export function readJson(filePath) {
  try {
    return JSON.parse(readFileSync(filePath, 'utf-8'));
  } catch {
    return null;
  }
}

export function readPackageJson(dir) {
  return readJson(join(dir, 'package.json'));
}

export function readDenoJson(dir) {
  for (const name of ['deno.json', 'deno.jsonc']) {
    const r = readJson(join(dir, name));
    if (r) return r;
  }
  return null;
}

// ── Package Utils ─────────────────────────────────────────────

export function getAllPackageNames(pkg) {
  if (!pkg) return [];
  return [
    ...Object.keys(pkg.dependencies || {}),
    ...Object.keys(pkg.devDependencies || {}),
  ];
}

/**
 * Reads the installed version of a package from node_modules.
 * Returns a parsed major version number or null.
 */
export function getInstalledMajorVersion(dir, packageName) {
  const pkgPath = join(dir, 'node_modules', packageName, 'package.json');
  const pkg = readJson(pkgPath);
  if (!pkg?.version) return null;
  const major = parseInt(pkg.version.split('.')[0], 10);
  return isNaN(major) ? null : major;
}

/**
 * Reads the declared version range of a dependency from package.json.
 * Strips semver prefixes (^, ~, >=) and returns the major number.
 */
export function getDeclaredMajorVersion(pkg, packageName) {
  const allDeps = { ...pkg?.dependencies, ...pkg?.devDependencies };
  const range = allDeps?.[packageName];
  if (!range) return null;
  const match = range.match(/(\d+)/);
  return match ? parseInt(match[1], 10) : null;
}

// ── Frontend File Scanner ────────────────────────────────────

export function hasWebFrontendFiles(dir, maxDepth = 3) {
  function scan(d, depth) {
    let entries;
    try { entries = readdirSync(d, { withFileTypes: true }); } catch { return false; }
    for (const entry of entries) {
      if (entry.isFile()) {
        if (entry.name.endsWith('.blade.php')) return true;
        const ext = extname(entry.name);
        if (ext && WEB_FRONTEND_EXTENSIONS.has(ext)) return true;
      } else if (entry.isDirectory() && depth < maxDepth) {
        if (SKIP_DIRS.has(entry.name) || entry.name.startsWith('.')) continue;
        if (scan(join(d, entry.name), depth + 1)) return true;
      }
    }
    return false;
  }
  return scan(dir, 0);
}

/**
 * Checks if the project root has files with given extensions (non-recursive).
 * Used for .csproj/.sln detection.
 */
function hasFilesWithExtension(dir, extensions) {
  const exts = new Set(extensions);
  try {
    return readdirSync(dir).some((f) => exts.has(extname(f)));
  } catch {
    return false;
  }
}

// ── Workspace Resolution ─────────────────────────────────────

function parsePnpmWorkspace(content) {
  const patterns = [];
  let inPkgs = false;
  for (const raw of content.split('\n')) {
    const line = raw.trim();
    if (line === 'packages:') { inPkgs = true; continue; }
    if (inPkgs) {
      if (line.startsWith('- ')) patterns.push(line.slice(2).trim().replace(/^['"]|['"]$/g, ''));
      else if (line && !line.startsWith('#')) break;
    }
  }
  return patterns;
}

function expandGlobs(dir, patterns) {
  const dirs = [];
  for (const pattern of patterns) {
    if (pattern.includes('*')) {
      const parent = join(dir, pattern.replace(/\/?\*.*$/, ''));
      try {
        for (const entry of readdirSync(parent, { withFileTypes: true })) {
          if (!entry.isDirectory() || SKIP_DIRS.has(entry.name) || entry.name.startsWith('.')) continue;
          const wsDir = join(parent, entry.name);
          if (existsSync(join(wsDir, 'package.json')) || existsSync(join(wsDir, 'deno.json'))) {
            dirs.push(wsDir);
          }
        }
      } catch {}
    } else {
      const wsDir = join(dir, pattern);
      if (existsSync(join(wsDir, 'package.json')) || existsSync(join(wsDir, 'deno.json'))) {
        dirs.push(wsDir);
      }
    }
  }
  return dirs;
}

function resolveWorkspaces(dir, pkg) {
  const pnpmPath = join(dir, 'pnpm-workspace.yaml');
  if (existsSync(pnpmPath)) {
    try {
      const patterns = parsePnpmWorkspace(readFileSync(pnpmPath, 'utf-8'));
      if (patterns.length) return expandGlobs(dir, patterns).filter((d) => resolve(d) !== resolve(dir));
    } catch {}
  }
  const ws = pkg?.workspaces;
  const patterns = Array.isArray(ws) ? ws : Array.isArray(ws?.packages) ? ws.packages : null;
  if (patterns?.length) return expandGlobs(dir, patterns).filter((d) => resolve(d) !== resolve(dir));
  return [];
}

// ── Stack Detection ───────────────────────────────────────────

function detectInDir(dir, preloadedPkg) {
  const pkg = preloadedPkg !== undefined ? preloadedPkg : readPackageJson(dir);
  const pkgNames = new Set(getAllPackageNames(pkg));
  const fileCache = new Map();
  const existsCache = new Map();

  function cachedRead(p) {
    if (fileCache.has(p)) return fileCache.get(p);
    let c = null;
    try { c = readFileSync(p, 'utf-8'); } catch {}
    fileCache.set(p, c);
    if (c !== null) existsCache.set(p, true);
    return c;
  }

  function cachedExists(p) {
    if (existsCache.has(p)) return existsCache.get(p);
    const r = existsSync(p);
    existsCache.set(p, r);
    return r;
  }

  const detected = [];

  for (const tech of SKILLS_MAP) {
    let found = false;

    if (!found && tech.detect.packages?.length) {
      found = tech.detect.packages.some((p) => pkgNames.has(p));
    }
    if (!found && tech.detect.packagePatterns?.length) {
      found = tech.detect.packagePatterns.some((re) => [...pkgNames].some((p) => re.test(p)));
    }
    if (!found && tech.detect.configFiles?.length) {
      found = tech.detect.configFiles.some((f) => cachedExists(join(dir, f)));
    }
    if (!found && tech.detect.configFileContent?.patterns?.length) {
      const cfg = tech.detect.configFileContent;
      for (const fp of (cfg.files || []).map((f) => join(dir, f))) {
        const content = cachedRead(fp);
        if (content && cfg.patterns.some((p) => content.includes(p))) { found = true; break; }
      }
    }
    if (!found && tech.detect.fileExtensions?.length) {
      found = hasFilesWithExtension(dir, tech.detect.fileExtensions);
    }

    if (found) {
      // Angular version filtering
      if (tech.id.startsWith('angular')) {
        const major = getInstalledMajorVersion(dir, '@angular/core')
          ?? getDeclaredMajorVersion(pkg, '@angular/core')
          ?? 999; // unknown → assume modern

        if (tech.minAngularVersion && major < tech.minAngularVersion) continue;
        if (tech.maxAngularVersion && major > tech.maxAngularVersion) continue;
      }

      // AG Grid version filtering
      if (tech.id.startsWith('ag-grid')) {
        const major = getInstalledMajorVersion(dir, 'ag-grid-community')
          ?? getInstalledMajorVersion(dir, '@ag-grid-community/core')
          ?? getDeclaredMajorVersion(pkg, 'ag-grid-community')
          ?? getDeclaredMajorVersion(pkg, '@ag-grid-community/core')
          ?? 999;

        if (tech.id === 'ag-grid-v32' && major < 32) continue;
        if (tech.id === 'ag-grid-legacy' && major >= 32) continue;
      }

      detected.push(tech);
    }
  }

  const isFrontendByPackages = [...pkgNames].some((p) => FRONTEND_PACKAGES.has(p));
  const isFrontendByFiles = isFrontendByPackages ? false : hasWebFrontendFiles(dir);

  return { detected, isFrontend: isFrontendByPackages || isFrontendByFiles };
}

// ── Main Entry ────────────────────────────────────────────────

/**
 * Detects technologies in projectDir (including workspaces).
 * Returns { detected: TechEntry[], isFrontend: boolean }
 */
export function detectTechnologies(projectDir) {
  const pkg = readPackageJson(projectDir);
  const root = detectInDir(projectDir, pkg);

  const seen = new Map(root.detected.map((t) => [t.id, t]));
  let isFrontend = root.isFrontend;

  for (const wsDir of resolveWorkspaces(projectDir, pkg)) {
    const ws = detectInDir(wsDir);
    for (const tech of ws.detected) {
      if (!seen.has(tech.id)) seen.set(tech.id, tech);
    }
    if (ws.isFrontend) isFrontend = true;
  }

  return { detected: [...seen.values()], isFrontend };
}

/**
 * Collects suggested MCP skills from detected technologies + universal skills.
 * Each entry: { skillName, version, sources, category }
 *
 * category: 'stack' | 'universal' | 'sdd'
 */
export function collectSkills({ detected, installedNames = new Set() }) {
  const map = new Map();

  function add(skillName, repo, subdir, source, category) {
    const key = skillName;
    if (!map.has(key)) {
      map.set(key, {
        skillName,
        repo,
        subdir,
        sources: [source],
        category,
        alreadyInstalled: installedNames.has(`${skillName}.md`),
      });
    } else {
      const entry = map.get(key);
      if (!entry.sources.includes(source)) entry.sources.push(source);
    }
  }

  // Stack-specific skills
  for (const tech of detected) {
    for (const s of tech.mcpSkills || []) {
      add(s.skillName, s.repo, s.subdir, tech.name, 'stack');
    }
  }

  // Universal skills
  for (const s of UNIVERSAL_SKILLS) {
    add(s.skillName, s.repo, s.subdir, 'Universal', 'universal');
  }

  return [...map.values()];
}

/**
 * Reads the set of already-installed skill file names from .claude/commands/.
 */
export function getInstalledSkillNames(projectDir) {
  const commandsDir = join(projectDir, '.claude', 'commands');
  try {
    return new Set(readdirSync(commandsDir).filter((f) => f.endsWith('.md')));
  } catch {
    return new Set();
  }
}

/**
 * Detects which agents are installed by checking known home dirs.
 */
export function detectInstalledAgents(home = homedir()) {
  return Object.values(AGENTS).filter((a) => existsSync(join(home, a.homeDir)));
}
