// @favelasquez/agentskills/src/skills-map.mjs
/**
 * SKILLS_MAP
 *
 * Maps detected stack technologies → skill suggestions.
 * Each entry:
 *   id        : unique tech identifier
 *   name      : human-readable label
 *   detect    : detection rules (packages | configFiles | configFileContent | packagePatterns)
 *   mcpSkills : array of { skillName, repo } to suggest when this tech is detected
 *              repo format: '{owner}/{repo}' (GitHub)
 *              fetched from: https://raw.githubusercontent.com/{repo}/HEAD/{skillName}/SKILL.md
 */
export const SKILLS_MAP = [
  // ── Angular ──────────────────────────────────────────────────
  // Angular version detection handled in collectSkills (reads @angular/core version)
  {
    id: 'angular-modern',
    name: 'Angular (v13+)',
    detect: {
      packages: ['@angular/core'],
      configFiles: ['angular.json'],
    },
    // Only suggest when Angular >= 13 — resolved dynamically in collectSkills
    minAngularVersion: 13,
    mcpSkills: [
      { skillName: 'angular-developer', repo: 'angular/skills' },
    ],
  },
  {
    id: 'angular-mid',
    name: 'Angular (v6–v12)',
    detect: {
      packages: ['@angular/core'],
      configFiles: ['angular.json'],
    },
    minAngularVersion: 6,
    maxAngularVersion: 12,
    mcpSkills: [
      { skillName: 'angular-developer', repo: 'angular/skills' },
    ],
  },
  {
    id: 'angular-legacy',
    name: 'Angular (legacy < v6)',
    detect: {
      packages: ['@angular/core'],
      configFiles: ['angular.json'],
    },
    maxAngularVersion: 5,
    mcpSkills: [
      { skillName: 'angular-developer', repo: 'angular/skills' },
    ],
  },
  // ── Laravel (PHP) ────────────────────────────────────────────
  {
    id: 'laravel',
    name: 'Laravel',
    detect: {
      configFiles: ['artisan', 'composer.json'],
      configFileContent: {
        files: ['composer.json'],
        patterns: ['"laravel/framework"'],
      },
    },
    mcpSkills: [
      { skillName: 'laravel-audit', version: 'v1' },
      { skillName: 'stack-detector', version: 'v1' },
    ],
  },
  // ── Python / FastAPI ─────────────────────────────────────────
  {
    id: 'fastapi',
    name: 'FastAPI',
    detect: {
      configFiles: ['requirements.txt', 'pyproject.toml'],
      configFileContent: {
        files: ['requirements.txt', 'pyproject.toml'],
        patterns: ['fastapi', 'FastAPI'],
      },
    },
    mcpSkills: [
      { skillName: 'api-design', repo: 'wdm0006/python-skills', subdir: 'skills' },
      { skillName: 'testing-strategy', repo: 'wdm0006/python-skills', subdir: 'skills' },
      { skillName: 'security-audit', repo: 'wdm0006/python-skills', subdir: 'skills' },
    ],
  },
  // ── C# / .NET ────────────────────────────────────────────────
  {
    id: 'csharp',
    name: 'C# / .NET',
    detect: {
      configFiles: [],
      configFileContent: {
        files: [],
        patterns: [],
        scanGradleLayout: false,
      },
      fileExtensions: ['.csproj', '.sln'],
    },
    mcpSkills: [
      { skillName: 'csharp-ef-audit', version: 'v1' },
      { skillName: 'stack-detector', version: 'v1' },
    ],
  },
  // ── AG Grid ──────────────────────────────────────────────────
  {
    id: 'ag-grid-v32',
    name: 'AG Grid (v32+)',
    detect: {
      packages: ['ag-grid-community', '@ag-grid-community/core', 'ag-grid-angular', 'ag-grid-react'],
    },
    minVersion: '32.0.0',
    mcpSkills: [
      { skillName: 'ag-grid', version: 'v32' },
    ],
  },
  {
    id: 'ag-grid-legacy',
    name: 'AG Grid (< v32)',
    detect: {
      packages: ['ag-grid-community', '@ag-grid-community/core', 'ag-grid-angular', 'ag-grid-react'],
    },
    maxVersion: '31.9.9',
    mcpSkills: [
      { skillName: 'ag-grid', version: 'v18' },
    ],
  },
  // ── Deepgram ─────────────────────────────────────────────────
  {
    id: 'deepgram',
    name: 'Deepgram',
    detect: {
      packages: ['@deepgram/sdk', 'deepgram'],
    },
    mcpSkills: [
      { skillName: 'deepgram-expert', version: 'v1' },
    ],
  },
  // ── Twilio ───────────────────────────────────────────────────
  {
    id: 'twilio',
    name: 'Twilio',
    detect: {
      packages: ['twilio', '@twilio/conversations', '@twilio/voice-sdk'],
    },
    mcpSkills: [
      { skillName: 'twilio-expert', version: 'v1' },
    ],
  },
];

/**
 * UNIVERSAL_SKILLS
 *
 * Always suggested regardless of detected stack.
 * These are workflow/tooling skills useful for any project.
 */
// Universal skills — add repo field when URLs are ready
export const UNIVERSAL_SKILLS = [];

/**
 * SDD_SKILLS
 *
 * Spec-Driven Development workflow skills — suggested as a group.
 */
export const SDD_SKILLS = [
  { skillName: 'sdd-init', version: 'v1' },
  { skillName: 'sdd-explore', version: 'v1' },
  { skillName: 'sdd-propose', version: 'v1' },
  { skillName: 'sdd-spec', version: 'v1' },
  { skillName: 'sdd-design', version: 'v1' },
  { skillName: 'sdd-tasks', version: 'v1' },
  { skillName: 'sdd-apply', version: 'v1' },
  { skillName: 'sdd-verify', version: 'v1' },
  { skillName: 'sdd-archive', version: 'v1' },
];

/**
 * File extensions that indicate a web frontend project.
 */
export const WEB_FRONTEND_EXTENSIONS = new Set([
  '.html', '.htm', '.css', '.scss', '.sass', '.less',
  '.vue', '.svelte', '.jsx', '.tsx',
]);

/**
 * Packages that indicate a web frontend project.
 */
export const FRONTEND_PACKAGES = new Set([
  'react', 'react-dom', 'vue', 'svelte', '@sveltejs/kit',
  'angular', '@angular/core', 'astro', 'solid-js',
]);
