// @favelasquez/agentskills/src/ui.mjs
import * as clack from '@clack/prompts';

// ── Helpers ──────────────────────────────────────────────────

const c = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  dim: '\x1b[2m',
  cyan: '\x1b[36m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  red: '\x1b[31m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
};

const fmt = {
  bold: (s) => `${c.bold}${s}${c.reset}`,
  dim: (s) => `${c.dim}${s}${c.reset}`,
  cyan: (s) => `${c.cyan}${s}${c.reset}`,
  green: (s) => `${c.green}${s}${c.reset}`,
  yellow: (s) => `${c.yellow}${s}${c.reset}`,
  red: (s) => `${c.red}${s}${c.reset}`,
  blue: (s) => `${c.blue}${s}${c.reset}`,
  magenta: (s) => `${c.magenta}${s}${c.reset}`,
};

// ── Intro ─────────────────────────────────────────────────────

export function printIntro() {
  clack.intro(fmt.bold(fmt.cyan(' autoskills ')) + fmt.dim('— AI skills for your stack'));
}

export function printOutro(installed, failed) {
  if (failed > 0) {
    clack.outro(`${fmt.green(`✔ ${installed} installed`)}  ${fmt.red(`✘ ${failed} failed`)}`);
  } else if (installed > 0) {
    clack.outro(fmt.green(`✔ ${installed} skill${installed !== 1 ? 's' : ''} installed successfully`));
  } else {
    clack.outro(fmt.dim('Nothing to install. All good!'));
  }
}

// ── Detection Summary ────────────────────────────────────────

export function printDetected(detected, combos, isFrontend) {
  if (!detected.length && !isFrontend) {
    clack.log.warn('No recognizable stack detected in this directory.');
    return;
  }

  const techs = detected.map((t) => fmt.cyan(t.name));
  if (isFrontend) techs.push(fmt.cyan('Frontend'));

  clack.log.info(`Detected: ${techs.join(fmt.dim(', '))}`);

  if (combos.length > 0) {
    clack.log.info(`Combos: ${combos.map((c) => fmt.magenta(c.name)).join(', ')}`);
  }
}

// ── Agent Selection ──────────────────────────────────────────

/**
 * Ask the user which agents to target (when no --flags were passed).
 * Shows only agents that appear to be installed; always includes "all".
 */
export async function promptAgents(installedAgents, allAgents) {
  const options = [
    { value: '__all__', label: fmt.bold('All supported agents') },
    ...installedAgents.map((a) => ({
      value: a.flag,
      label: `${a.name} ${fmt.dim('(detected)')}`,
    })),
    ...allAgents
      .filter((a) => !installedAgents.find((i) => i.flag === a.flag))
      .map((a) => ({
        value: a.flag,
        label: fmt.dim(a.name),
      })),
  ];

  const selected = await clack.multiselect({
    message: 'Which agents should receive the skills?',
    options,
    required: true,
  });

  if (clack.isCancel(selected)) {
    clack.cancel('Cancelled.');
    process.exit(0);
  }

  if (selected.includes('__all__')) {
    return allAgents.map((a) => a.flag);
  }
  return selected;
}

// ── Skill Selection ──────────────────────────────────────────

/**
 * Present suggested skills as checkboxes. User picks which ones to install.
 * Pre-selects all non-installed skills by default.
 */
export async function promptSkills(suggestions) {
  if (!suggestions.length) {
    clack.log.warn('No skills found for the detected stack.');
    return [];
  }

  // Group: stack-specific first, then universal
  const stack = suggestions.filter((s) => s.category === 'stack');
  const universal = suggestions.filter((s) => s.category !== 'stack');
  const ordered = [...stack, ...universal];

  const options = ordered.map((s) => {
    const key = s.skillName;
    const label = s.alreadyInstalled
      ? `${fmt.dim(s.displayName || s.skillName)} ${fmt.green('✓ installed')}`
      : `${fmt.bold(s.displayName || s.skillName)} ${fmt.dim(`← ${s.sources.join(', ')}`)}`;
    return {
      value: key,
      label,
      hint: s.description || undefined,
    };
  });

  const selected = await clack.multiselect({
    message: `Select skills to install ${fmt.dim(`(${suggestions.length} suggested)`)}`,
    options,
    initialValues: ordered.filter((s) => !s.alreadyInstalled && s.category === 'stack').map((s) => s.skillName),
    required: false,
  });

  if (clack.isCancel(selected)) {
    clack.cancel('Cancelled.');
    process.exit(0);
  }

  return selected;
}

// ── Dry-run Summary ──────────────────────────────────────────

export function printDryRun(selectedSkills, agentFlags) {
  clack.log.info(fmt.yellow('Dry-run — nothing will be installed.'));

  if (!selectedSkills.length) {
    clack.log.message('No skills selected.');
    return;
  }

  const agentLabel = agentFlags.length ? agentFlags.join(', ') : 'all agents';
  clack.log.message(
    `Would install ${fmt.bold(String(selectedSkills.length))} skill(s) for ${fmt.cyan(agentLabel)}:\n` +
    selectedSkills.map((s) => `  ${fmt.dim('→')} ${s}`).join('\n'),
  );
}

// ── Install Progress ─────────────────────────────────────────

/**
 * Wraps an async install function with a clack spinner per skill.
 * Returns { installed, failed, errors }.
 */
export async function runWithProgress(skills, agentIds, installFn) {
  let installed = 0;
  let failed = 0;
  const errors = [];

  const spinner = clack.spinner();

  for (const skill of skills) {
    const label = typeof skill === 'string' ? skill : skill.skillName;
    spinner.start(`Installing ${fmt.cyan(label)}…`);
    const result = await installFn(skill, agentIds);
    if (result.success) {
      spinner.stop(`${fmt.green('✔')} ${label}`);
      installed++;
    } else {
      spinner.stop(`${fmt.red('✘')} ${label}`);
      failed++;
      errors.push({ skill: label, output: result.output });
    }
  }

  if (errors.length > 0) {
    clack.log.error('Some skills failed to install:');
    for (const e of errors) {
      clack.log.message(`  ${fmt.red(e.skill)}\n  ${fmt.dim(e.output.trim())}`);
    }
  }

  return { installed, failed, errors };
}

// ── No-stack fallback ────────────────────────────────────────

export async function promptManualSkill() {
  const skill = await clack.text({
    message: 'Enter a skill path to install (e.g. owner/repo/skill-name):',
    placeholder: 'owner/repo/skill-name',
    validate: (v) => (v.includes('/') ? undefined : 'Must be in owner/repo format'),
  });

  if (clack.isCancel(skill)) {
    clack.cancel('Cancelled.');
    process.exit(0);
  }

  return skill;
}
